File: Sorts/RadixSort.py
"""
Brian Taylor Vann
github.com/taylor-vann

Description:
- Module to Radix sort a list

Requirements:
- Assumes whole numbers (pos ints: 0, 1, 2, ...)

Methods:
- radix_sort
- bucketish_sort
- get_max
"""


def radix_sort(a, base = 10):
    if len(a) <= 1:
        return a

    bckt = a
    exp = base
    mod = 1
    max_num = get_max(a)

    while max_num//mod > 0:
        bckt = bucketish_sort(bckt, exp, mod)
        exp *= base
        mod *= base

    return bckt


def bucketish_sort(a, exp, mod, base = 10):
    bckt = []

    for i in range(base):
        bckt.append([])

    coll = []

    for val in a:
        index = (val % exp)//mod
        bckt[index].append(val)

    for var in bckt:
        coll += var

    return coll


def get_max(a):
    if len(a) == 0:
        return None

    n = a[0]

    for v in a:
        if n < v:
            n = v

    return n

File: ECMAScript/Sorts/RadixSort.js
/*
Brian Taylor Vann
github.com/taylor-vann

Description:
- Module to Radix sort a list

Requirements:
- Assumes whole numbers
- Least Significant Digit (LSD) Radix Sort

Methods:
- radix_sort
- bucketish_sort
*/

function radixSort(a, base = 10) {
  if(a.length < 2) {
    return a;
  }

  var srtd = a.slice();
  var pos = base;
  var mod = 1;
  var maxNum = Math.max(...a);

  while(Math.floor(maxNum / mod) > 0) {
    srtd = digitSort(srtd, pos, mod, base);
    pos *= base;
    mod *= base;
  }

  return srtd;
}


function digitSort(a, pos, mod, base) {
  var bckt = [];

  for(j = 0; j < base; j++) {
    bckt.push([]);
  }

  for(k = 0; k < a.length; k++) {
    var indx = Math.floor((a[k] % pos) / mod);
    bckt[indx].push(a[k])
  }

  var coll = [];

  return coll.concat(...bckt);
}
